examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from lightning.fabric.fabric import (
+    _PLUGIN_INPUT,
+    _PRECISION_INPUT,
+    _unwrap_objects,
+    Accelerator,
+    Fabric,
+    Logger,
+    Strategy,
+)
+from lightning.fabric.strategies.fsdp import FSDPStrategy
+from lightning.fabric.utilities.rank_zero import rank_zero_warn
+from lightning.fabric.utilities.types import LRScheduler, Optimizable
+from lightning.pytorch.core.module import LightningModule
+
+
+class Trainer:
+    def __init__(
+        self,
+        accelerator: Union[str, Accelerator] = "auto",
+        strategy: Union[str, Strategy] = "auto",
+        devices: Union[List[int], str, int] = "auto",
+        precision: _PRECISION_INPUT = "32-true",
+        plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
+        callbacks: Optional[Union[List[Any], Any]] = None,
+        loggers: Optional[Union[Logger, List[Logger]]] = None,
+        max_epochs: Optional[int] = 1000,
+        max_steps: Optional[int] = None,
+        grad_accum_steps: int = 1,
+        limit_train_batches: Union[int, float] = float("inf"),
+        limit_val_batches: Union[int, float] = float("inf"),
+        val_sanity_check: int = 2,
+        validation_frequency: int = 1,
+        use_distributed_sampler: bool = True,
+        checkpoint_dir: str = "./checkpoints",
+        checkpoint_frequency: int = 1,
+    ) -> None:
+
+        self.fabric = Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+            plugins=plugins,
+            callbacks=callbacks,
+            loggers=loggers,
+        )
+        self.global_step = 0
+        self.grad_accum_steps: int = grad_accum_steps
+        self.current_epoch = 0
+
+        self.max_epochs = max_epochs
+        self.max_steps = max_steps
+        self.should_stop = False
+
+        # ensures limit_X_batches is either int or inf
+        if not isinstance(limit_train_batches, int):
+            assert limit_train_batches == float("inf")
+
+        if not isinstance(limit_val_batches, int):
+            assert limit_val_batches == float("inf")
+
+        self.limit_train_batches = limit_train_batches
+        self.limit_val_batches = limit_val_batches
+        self.val_sanity_check = val_sanity_check
+        self.validation_frequency = validation_frequency
+        self.use_distributed_sampler = use_distributed_sampler
+        self._current_train_return: Union[torch.Tensor, Mapping[str, Any]] = {}
+        self._current_val_return: Optional[Union[torch.Tensor, Mapping[str, Any]]] = {}
+
+        self.checkpoint_dir = checkpoint_dir
+        self.checkpoint_frequency = checkpoint_frequency
+
+    def fit(
+        self, model: LightningModule, train_loader: DataLoader, val_loader: DataLoader, ckpt_path: Optional[str] = None
+    ):
+        if self.fabric.is_global_zero:
+            self.fabric.call("prepare_data")
```

Comment:
```suggestion
        if self.fabric.local_rank == 0:
            self.fabric.call("prepare_data")
```

Reply:
prepare_data is only defined on LightningDataModule. Should we perhaps just drop this?

Reply:
yes, will remove this!